File: compiler/src/zinc/compiler/Ast.java
// Copyright 2026 victorybhg
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0

package zinc.compiler;

import java.util.List;

public final class Ast {

    private Ast() {}

    // --- Root ----------------------------------------------------------------

    public record Program(
        String sourceFile,
        PackageDecl pkg,           // nullable
        List<ImportDecl> imports,
        List<TopLevelDecl> decls,
        List<Stmt> stmts           // script mode top-level statements
    ) {}

    public record PackageDecl(String path) {}
    public record ImportDecl(String path, String alias) {}

    // --- Top-level declarations ----------------------------------------------

    public sealed interface TopLevelDecl permits
        ClassDecl, InterfaceDecl, FnDecl, DataClassDecl, EnumDecl, ConstDecl, SealedClassDecl {}

    public record ClassDecl(
        int line, String name, boolean isAbstract,
        List<String> typeParams, List<String> parents,
        List<FieldDecl> fields, List<CtorDecl> ctors,
        List<MethodDecl> methods, List<Annotation> annotations
    ) implements TopLevelDecl {}

    public record SealedClassDecl(
        int line, String name, List<String> parents,
        List<FieldDecl> fields, List<CtorDecl> ctors,
        List<MethodDecl> methods, List<DataClassDecl> variants,
        List<Annotation> annotations
    ) implements TopLevelDecl {}

    public record InterfaceDecl(
        int line, String name, List<MethodSig> methods
    ) implements TopLevelDecl {}

    public record DataClassDecl(
        int line, String name, List<String> typeParams,
        List<String> parents, List<FieldDecl> params,
        List<MethodDecl> methods
    ) implements TopLevelDecl {}

    public record EnumDecl(int line, String name, List<String> variants) implements TopLevelDecl {}

    public record ConstDecl(int line, String name, boolean isPub, TypeExpr type, Expr value) implements TopLevelDecl {}

    public record FnDecl(
        int line, String name, boolean isPub,
        List<String> typeParams, List<ParamDecl> params,
        TypeExpr returnType, BlockStmt body,
        List<Annotation> annotations
    ) implements TopLevelDecl, Stmt {}

    // --- Supporting declarations ---------------------------------------------

    public record MethodSig(String name, boolean isPub, List<ParamDecl> params, TypeExpr returnType) {}

    public record CtorDecl(List<ParamDecl> params, BlockStmt body, List<Expr> superArgs) {}

    public record MethodDecl(
        String name, boolean isPub, boolean isStatic, boolean isAbstract,
        List<ParamDecl> params, TypeExpr returnType,
        BlockStmt body, List<Annotation> annotations
    ) {}

    public record FieldDecl(
        String name, boolean isPub, boolean isReadonly, boolean isConst, boolean isInit,
        TypeExpr type, Expr defaultValue, List<Annotation> annotations
    ) {}

    public record ParamDecl(String name, TypeExpr type, Expr defaultValue, boolean isVariadic) {}

    public record NamedArg(String name, Expr value) {}

    public record Annotation(String name, List<String> args) {}

    // --- Type expressions ----------------------------------------------------

    public sealed interface TypeExpr permits
        SimpleType, GenericType, OptionalType, ArrayType, FuncType {}

    public record SimpleType(String name) implements TypeExpr {}
    public record GenericType(String name, List<TypeExpr> typeArgs) implements TypeExpr {}
    public record OptionalType(TypeExpr inner) implements TypeExpr {}
    public record ArrayType(TypeExpr elementType) implements TypeExpr {}
    public record FuncType(List<TypeExpr> params, TypeExpr returnType) implements TypeExpr {}

    // --- Statements ----------------------------------------------------------

    public sealed interface Stmt permits
        BlockStmt, VarStmt, TupleVarStmt, AssignStmt, ReturnStmt,
        IfStmt, ForStmt, WhileStmt, MatchStmt,
        BreakStmt, ContinueStmt, ExprStmt,
        ParallelForStmt, ConcurrentStmt, TimeoutStmt,
        WithStmt, DeferStmt, LockStmt, FnDecl {}

    public record BlockStmt(List<Stmt> stmts) implements Stmt {}

    public record VarStmt(
        int line, String name, TypeExpr type, Expr value,
        boolean isConst, OrHandler orHandler
    ) implements Stmt {}

    public record TupleVarStmt(int line, List<String> names, Expr value, OrHandler orHandler) implements Stmt {}

    public record AssignStmt(int line, Expr target, String op, Expr value, OrHandler orHandler) implements Stmt {}

    public record ReturnStmt(int line, Expr value) implements Stmt {}

    public record IfStmt(int line, Expr cond, BlockStmt then, Stmt elseStmt) implements Stmt {}

    public record ForStmt(
        int line,
        Stmt init, Expr cond, Stmt post,           // C-style
        boolean isRange, String indexVar, String item, Expr range, // range-style
        BlockStmt body
    ) implements Stmt {}

    public record WhileStmt(int line, Expr cond, BlockStmt body) implements Stmt {}

    public record MatchStmt(int line, Expr subject, List<MatchCase> cases) implements Stmt {}
    public record MatchCase(Expr pattern, BlockStmt body) {}

    public record BreakStmt() implements Stmt {}
    public record ContinueStmt() implements Stmt {}

    public record ExprStmt(int line, Expr expr, OrHandler orHandler) implements Stmt {}

    public record ParallelForStmt(
        int line, String item, String indexVar, Expr range,
        BlockStmt body, OrHandler orHandler, int max
    ) implements Stmt {}

    public record ConcurrentStmt(
        int line, List<Expr> tasks, boolean firstOnly,
        List<String> names, OrHandler orHandler
    ) implements Stmt {}

    public record TimeoutStmt(int line, Expr duration, BlockStmt body, OrHandler orHandler) implements Stmt {}

    public record WithStmt(int line, List<WithResource> resources, BlockStmt body) implements Stmt {}
    public record WithResource(String name, Expr value, OrHandler orHandler) {}

    public record DeferStmt(Expr expr) implements Stmt {}

    public record LockStmt(int line, Expr mutex, BlockStmt body) implements Stmt {}

    // --- Error handling ------------------------------------------------------

    public record OrHandler(BlockStmt body, List<OrMatchCase> matchCases, String matchVar) {}
    public record OrMatchCase(String type, BlockStmt body) {}

    // --- Expressions ---------------------------------------------------------

    public sealed interface Expr permits
        IntLit, FloatLit, StringLit, StringInterpLit, RawStringLit,
        BoolLit, NullLit, Ident, ThisExpr,
        BinaryExpr, UnaryExpr, CallExpr, SelectorExpr, SafeNavExpr,
        IndexExpr, SliceExpr, ListLit, MapLit,
        LambdaExpr, SpawnExpr, IfExpr, MatchExpr,
        RangeExpr, TupleLit, SpreadExpr, TypeAssertExpr,
        SuperCallExpr {}

    public record IntLit(String value) implements Expr {}
    public record FloatLit(String value) implements Expr {}
    public record StringLit(String value) implements Expr {}
    public record StringInterpLit(List<Expr> parts) implements Expr {}
    public record RawStringLit(String value) implements Expr {}
    public record BoolLit(boolean value) implements Expr {}
    public record NullLit() implements Expr {}
    public record Ident(String name) implements Expr {}
    public record ThisExpr() implements Expr {}

    public record BinaryExpr(Expr left, String op, Expr right) implements Expr {}
    public record UnaryExpr(String op, Expr operand) implements Expr {}

    public record CallExpr(
        Expr callee, List<Expr> args, List<NamedArg> namedArgs,
        List<String> typeArgs, boolean isNew
    ) implements Expr {}

    public record SelectorExpr(Expr object, String field) implements Expr {}
    public record SafeNavExpr(Expr object, String field, CallExpr call) implements Expr {}
    public record IndexExpr(Expr object, Expr index) implements Expr {}
    public record SliceExpr(Expr object, Expr low, Expr high) implements Expr {}

    public record ListLit(List<Expr> elements) implements Expr {}
    public record MapLit(List<Expr> keys, List<Expr> values) implements Expr {}

    public record LambdaExpr(List<ParamDecl> params, BlockStmt body) implements Expr {}

    public record SpawnExpr(int line, BlockStmt body, OrHandler orHandler) implements Expr {}

    public record IfExpr(Expr cond, Expr then, Expr elseExpr) implements Expr {}
    public record MatchExpr(Expr subject, List<MatchExprCase> cases) implements Expr {}
    public record MatchExprCase(Expr pattern, Expr value) {}

    public record RangeExpr(Expr start, Expr end, boolean inclusive) implements Expr {}
    public record TupleLit(List<Expr> elements) implements Expr {}
    public record SpreadExpr(Expr expr) implements Expr {}
    public record TypeAssertExpr(Expr object, String typeName, boolean isCheck) implements Expr {}
    public record SuperCallExpr(List<Expr> args) implements Expr {}
}

File: zinc-go-py/zinc/ast.py
"""Zinc AST node types.

Dataclasses mirroring zinc-go/internal/parser/ast.go for easy cross-reference.
Fields carry only the information codegen needs — no Line/Pos plumbing yet.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional, Union

# --- Top-level ---------------------------------------------------------------

@dataclass
class Program:
    package: Optional["PackageDecl"] = None
    imports: list["ImportDecl"] = field(default_factory=list)
    decls: list["TopLevelDecl"] = field(default_factory=list)
    stmts: list["Stmt"] = field(default_factory=list)      # script-mode top-level
    source_file: str = ""

@dataclass
class PackageDecl:
    path: str

@dataclass
class ImportDecl:
    path: str
    alias: str = ""

# --- Declarations ------------------------------------------------------------

@dataclass
class Annotation:
    name: str
    args: list[str] = field(default_factory=list)

@dataclass
class ParamDecl:
    name: str
    type: Optional["TypeExpr"] = None
    default: Optional["Expr"] = None
    variadic: bool = False
    is_const: bool = False

@dataclass
class FieldDecl:
    name: str
    type: Optional["TypeExpr"]
    default: Optional["Expr"] = None
    is_pub: bool = False
    is_readonly: bool = False
    is_const: bool = False
    is_init: bool = False
    is_static: bool = False
    annotations: list[Annotation] = field(default_factory=list)

@dataclass
class FnDecl:
    name: str
    params: list[ParamDecl] = field(default_factory=list)
    return_type: Optional["TypeExpr"] = None     # None = void
    body: Optional["BlockStmt"] = None
    expr_body: Optional["Expr"] = None           # single-expression body
    is_pub: bool = False
    type_params: list[str] = field(default_factory=list)
    annotations: list[Annotation] = field(default_factory=list)

@dataclass
class CtorDecl:
    params: list[ParamDecl]
    body: "BlockStmt"

@dataclass
class MethodDecl:
    name: str
    params: list[ParamDecl] = field(default_factory=list)
    return_type: Optional["TypeExpr"] = None
    body: Optional["BlockStmt"] = None
    expr_body: Optional["Expr"] = None
    is_pub: bool = False
    is_static: bool = False
    is_abstract: bool = False
    is_override: bool = False
    annotations: list[Annotation] = field(default_factory=list)

@dataclass
class MethodSig:
    name: str
    params: list[ParamDecl] = field(default_factory=list)
    return_type: Optional["TypeExpr"] = None
    is_pub: bool = False

@dataclass
class ClassDecl:
    name: str
    fields: list[FieldDecl] = field(default_factory=list)
    ctor: Optional[CtorDecl] = None
    ctors: list[CtorDecl] = field(default_factory=list)
    methods: list[MethodDecl] = field(default_factory=list)
    parents: list[str] = field(default_factory=list)
    type_params: list[str] = field(default_factory=list)
    is_abstract: bool = False
    is_sealed: bool = False
    variants: list["DataClassDecl"] = field(default_factory=list)
    annotations: list[Annotation] = field(default_factory=list)

@dataclass
class DataClassDecl:
    name: str
    params: list[FieldDecl] = field(default_factory=list)
    methods: list[MethodDecl] = field(default_factory=list)
    parents: list[str] = field(default_factory=list)
    type_params: list[str] = field(default_factory=list)

@dataclass
class InterfaceDecl:
    name: str
    methods: list[MethodSig] = field(default_factory=list)
    type_params: list[str] = field(default_factory=list)

@dataclass
class EnumDecl:
    name: str
    variants: list[str] = field(default_factory=list)

@dataclass
class ConstDecl:
    name: str
    value: "Expr"
    type: Optional["TypeExpr"] = None
    is_pub: bool = False

@dataclass
class TypeAliasDecl:
    name: str
    type: "TypeExpr"

@dataclass
class TestDecl:
    name: str
    body: "BlockStmt"

TopLevelDecl = Union[
    ClassDecl, DataClassDecl, InterfaceDecl, EnumDecl, FnDecl, ConstDecl,
    TypeAliasDecl, TestDecl
]

# --- Types -------------------------------------------------------------------

@dataclass
class SimpleType:
    name: str

@dataclass
class GenericType:
    name: str
    type_args: list["TypeExpr"] = field(default_factory=list)

@dataclass
class ArrayType:
    element_type: "TypeExpr"

@dataclass
class OptionalType:
    inner: "TypeExpr"

@dataclass
class FuncTypeExpr:
    params: list["TypeExpr"] = field(default_factory=list)
    return_type: Optional["TypeExpr"] = None

TypeExpr = Union[SimpleType, GenericType, ArrayType, OptionalType, FuncTypeExpr]

# --- Statements --------------------------------------------------------------

@dataclass
class BlockStmt:
    stmts: list["Stmt"] = field(default_factory=list)

@dataclass
class VarStmt:
    name: str
    value: Optional["Expr"] = None
    type: Optional[TypeExpr] = None
    is_const: bool = False
    or_handler: Optional["OrHandler"] = None

@dataclass
class TupleVarStmt:
    names: list[str]
    value: "Expr"
    or_handler: Optional["OrHandler"] = None

@dataclass
class AssignStmt:
    target: "Expr"
    op: str
    value: "Expr"
    or_handler: Optional["OrHandler"] = None

@dataclass
class ReturnStmt:
    value: Optional["Expr"] = None

@dataclass
class IfStmt:
    cond: "Expr"
    then: BlockStmt
    else_: Optional["Stmt"] = None       # BlockStmt or IfStmt

@dataclass
class ForStmt:
    body: BlockStmt
    # C-style:
    init: Optional["Stmt"] = None
    cond: Optional["Expr"] = None
    post: Optional["Stmt"] = None
    # Range-style:
    is_range: bool = False
    index_var: str = ""
    item: str = ""
    range_expr: Optional["Expr"] = None

@dataclass
class WhileStmt:
    cond: "Expr"
    body: BlockStmt

@dataclass
class PrintStmt:
    value: "Expr"

@dataclass
class ExprStmt:
    expr: "Expr"
    or_handler: Optional["OrHandler"] = None

@dataclass
class BreakStmt:
    pass

@dataclass
class ContinueStmt:
    pass

@dataclass
class ThrowStmt:
    value: "Expr"

@dataclass
class DeferStmt:
    expr: "Expr"

@dataclass
class AssertStmt:
    cond: "Expr"
    message: Optional["Expr"] = None

@dataclass
class CatchClause:
    exception_type: Optional[TypeExpr]
    var_name: str
    body: BlockStmt

@dataclass
class TryStmt:
    body: BlockStmt
    catches: list[CatchClause] = field(default_factory=list)
    finally_: Optional[BlockStmt] = None

@dataclass
class WithResource:
    name: str
    value: "Expr"
    or_handler: Optional["OrHandler"] = None

@dataclass
class WithStmt:
    resources: list[WithResource]
    body: BlockStmt

@dataclass
class MatchCase:
    pattern: Optional["Expr"]          # None = wildcard
    body: BlockStmt

@dataclass
class MatchStmt:
    subject: "Expr"
    cases: list[MatchCase] = field(default_factory=list)

@dataclass
class ParallelForStmt:
    item: str
    range_expr: "Expr"
    body: BlockStmt
    index_var: str = ""
    max: Optional["Expr"] = None
    or_handler: Optional["OrHandler"] = None

@dataclass
class ConcurrentStmt:
    tasks: list["Expr"]
    first_only: bool = False
    names: list[str] = field(default_factory=list)
    or_handler: Optional["OrHandler"] = None

@dataclass
class TimeoutStmt:
    duration: "Expr"
    body: BlockStmt
    or_handler: Optional["OrHandler"] = None

@dataclass
class SpawnStmt:
    body: BlockStmt
    or_handler: Optional["OrHandler"] = None

@dataclass
class OrHandler:
    body: Optional[BlockStmt] = None
    default_value: Optional["Expr"] = None
    match_var: str = ""
    match_cases: list["OrMatchCase"] = field(default_factory=list)

@dataclass
class OrMatchCase:
    type_name: str
    body: BlockStmt

Stmt = Union[
    BlockStmt, VarStmt, TupleVarStmt, AssignStmt, ReturnStmt, IfStmt, ForStmt,
    WhileStmt, PrintStmt, ExprStmt, BreakStmt, ContinueStmt, ThrowStmt,
    DeferStmt, AssertStmt, TryStmt, WithStmt, MatchStmt, ParallelForStmt,
    ConcurrentStmt, TimeoutStmt, SpawnStmt, FnDecl,
]

# --- Expressions -------------------------------------------------------------

@dataclass
class BinaryExpr:
    left: "Expr"
    op: str
    right: "Expr"

@dataclass
class UnaryExpr:
    op: str
    operand: "Expr"

@dataclass
class NamedArg:
    name: str
    value: "Expr"

@dataclass
class StructFieldArg:
    name: str
    value: "Expr"

@dataclass
class CallExpr:
    callee: "Expr"
    args: list["Expr"] = field(default_factory=list)
    named_args: list[NamedArg] = field(default_factory=list)
    struct_field_args: list[StructFieldArg] = field(default_factory=list)
    type_args: list[TypeExpr] = field(default_factory=list)
    is_new: bool = False

@dataclass
class SelectorExpr:
    object: "Expr"
    field: str

@dataclass
class SafeNavExpr:
    object: "Expr"
    field: str
    call: Optional[CallExpr] = None

@dataclass
class PropagateExpr:
    inner: "Expr"

@dataclass
class IndexExpr:
    object: "Expr"
    index: "Expr"

@dataclass
class SliceExpr:
    object: "Expr"
    low: Optional["Expr"]
    high: Optional["Expr"]

@dataclass
class Ident:
    name: str

@dataclass
class IntLit:
    value: str

@dataclass
class FloatLit:
    value: str

@dataclass
class StringLit:
    value: str

@dataclass
class StringInterpLit:
    parts: list["Expr"]                # alternating StringLit and Expr

@dataclass
class RawStringLit:
    value: str

@dataclass
class BoolLit:
    value: bool

@dataclass
class NullLit:
    pass

@dataclass
class ThisExpr:
    pass

@dataclass
class SuperCallExpr:
    args: list["Expr"] = field(default_factory=list)

@dataclass
class ListLit:
    elements: list["Expr"] = field(default_factory=list)
    explicit_type: Optional[TypeExpr] = None

@dataclass
class MapLit:
    keys: list["Expr"] = field(default_factory=list)
    values: list["Expr"] = field(default_factory=list)
    explicit_type: Optional[TypeExpr] = None

@dataclass
class TupleLit:
    elements: list["Expr"] = field(default_factory=list)

@dataclass
class TypeAssertExpr:
    object: "Expr"
    type_name: str
    is_check: bool                     # True for `is`, False for `as`

@dataclass
class SpawnExpr:
    body: BlockStmt
    or_handler: Optional[OrHandler] = None

@dataclass
class SpreadExpr:
    expr: "Expr"

@dataclass
class LambdaExpr:
    params: list[ParamDecl]
    body: Optional[BlockStmt] = None
    expr: Optional["Expr"] = None
    return_type: Optional[TypeExpr] = None

@dataclass
class IfExpr:
    cond: "Expr"
    then: "Expr"
    else_: "Expr"

@dataclass
class MatchExprCase:
    pattern: Optional["Expr"]
    value: "Expr"

@dataclass
class MatchExpr:
    subject: "Expr"
    cases: list[MatchExprCase] = field(default_factory=list)

@dataclass
class RangeExpr:
    start: "Expr"
    end: "Expr"
    inclusive: bool = False

@dataclass
class CapacityExpr:
    collection_type: GenericType
    capacity: "Expr"

@dataclass
class SizedArrayExpr:
    element_type: str
    size: "Expr"

Expr = Union[
    BinaryExpr, UnaryExpr, CallExpr, SelectorExpr, SafeNavExpr, PropagateExpr,
    IndexExpr, SliceExpr, Ident, IntLit, FloatLit, StringLit, StringInterpLit,
    RawStringLit, BoolLit, NullLit, ThisExpr, SuperCallExpr, ListLit, MapLit,
    TupleLit, TypeAssertExpr, SpawnExpr, SpreadExpr, LambdaExpr, IfExpr,
    MatchExpr, RangeExpr, CapacityExpr, SizedArrayExpr,
]
